### PR TITLE
[dynamo] Remove uses of `raise unimplemented`

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122137
* #119430
* #119727
* __->__ #122136

`unimplemented` is a function that raises an error, so
`raise unimplemented(...)` never reaches the `raise`.
Another related issue is that `raise unimplemented(...) from e`
doesn't attach the exception cause correctly. I fix this by adding
a `from_exc` argument to `unimplemented`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_dynamo/exc.py [hunk 1]
```diff
@@ -182,11 +182,13 @@ def unimplemented_with_warning(e: Exception, code, msg: str) -> NoReturn:
...
     graph_breaks_log.debug("%s", graph_break_msg)
     log.warning(msg)
-    raise unimplemented(msg) from e
+    unimplemented(msg, from_exc=e)
 
 
-def unimplemented(msg: str) -> NoReturn:
+def unimplemented(msg: str, *, from_exc: Optional[Exception] = None) -> NoReturn:
     assert msg != os.environ.get("BREAK", False)
+    if from_exc is not None:
```
Comment (Skylion007) [inline torch/_dynamo/exc.py:190]: What if people want to raise from None to suppress the automatic exception chaining? You might need to use a sentinel value here instead.

**Review by lezcano — APPROVED** [review]
Nice. Feel free to merge once the linter error is fixed or skipped

**Pull request closed without merging**